### PR TITLE
🏌️ DAPO loss type

### PR DESCRIPTION
**Pull request by qgallouedec**
This PR introduces the true DAPO aggregation method for the loss. Previously, we were only able to achieve partial equivalence, because the normalizer was local and not accumulated. In other words, you could only achieve DAPO with gas=1 and num GPUs=1.

I checked that this new loss type is working and correct with both gradient accumulation and distributed training!

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -258,7 +263,9 @@ def shuffle_sequence_dict(seq_dict: dict[str, Optional[Sequence]]) -> dict[str,
     def permute(v: Optional[Sequence]) -> Optional[Sequence]:
         if v is None:
             return None
-        if isinstance(v, torch.Tensor):
+        if isinstance(v, torch.Tensor) and v.ndim == 0:
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:266]: this is for `"num_items_in_batch"` that is a single float that needs to be copied in every chunks

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1894,10 +1909,16 @@ def _compute_loss(self, model, inputs):
 
         if self.loss_type == "grpo":
             loss = ((per_token_loss * completion_mask).sum(-1) / completion_mask.sum(-1).clamp(min=1.0)).mean()
+            loss = loss / self.current_gradient_accumulation_steps
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1912]: now that we don't rely on trainer to scale the reward, we need to do it here.

**Review by shirinyamani — APPROVED** [review]
I watched you on X doing it, perfectly LGTM!

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -700,6 +707,12 @@ def __init__(
             processing_class=processing_class,
             callbacks=callbacks,
             optimizers=optimizers,
+            # In Trainer, `training_step` scales the loss by `gradient_accumulation_steps` only if `compute_loss_func`
+            # is None. For DAPO, loss scaling instead depends on the total number of completions tokens across the
+            # global accumulated batch. To control scaling ourselves, we must disable Trainer’s built-in scaling. The
+            # simplest (though a bit hacky) way is to set `compute_loss_func` to any non-None value, which bypasses
+            # that behavior without rewriting `training_step`.
+            compute_loss_func="non-None value to disable scaling",
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:715]: cc @SunMarc, idk if you'd have done it this way

Comment (SunMarc) [reply trl/trainer/grpo_trainer.py:715]: yeah it's a bit hacky but i guess that one of the simplest solution. Maybe you can also do it with modifying `model_accepts_loss_kwargs` and `num_items_in_batches`
